### PR TITLE
Rewrite of the record layer and MAC text

### PR DESCRIPTION
**Pull request by hannestschofenig**
Text to address the discussion in https://github.com/tlswg/dtls-conn-id/pull/35

### CODE REVIEW

**Comment by hannestschofenig** [comment]
Completely re-wrote PR #35

File: draft-ietf-tls-dtls-connection-id.md [hunk 1]
```diff
@@ -134,34 +133,48 @@ indicates that the server will send with the client's CID but does not
...
 Implementations, which want to use variable-length CIDs, are responsible
 for constructing the CID in such a way that its length can be determined
-on reception. Note that such implementations must still be able to send
-connection identifiers of different length to other parties.
-
-Note that it is not possible to parse the records without knowing how 
-long the CID is.
+on reception. Such implementations must still be able to send
+CIDs of different length to other parties. Is not possible to parse 
+the records without knowing how long the CID is.
```
Comment (boaks) [inline draft-ietf-tls-dtls-connection-id.md]: I'm not sure, why "Is not possible to parse the records without knowing how long the CID is." is kept. I tried to address my concerns in #27 . But this may be considered as independent from this PR.

**Comment by boaks** [comment]
Great work! So alternative I made it, right?

File: draft-ietf-tls-dtls-connection-id.md (hunk 1 above)
Comment (hannestschofenig) [reply draft-ietf-tls-dtls-connection-id.md]: Maybe it would be better to say that there is no CID length information included in the record itself. I think this would clarify the issue. Do you agree?

Comment (boaks) [reply draft-ietf-tls-dtls-connection-id.md]: Yes, that sounds good.

**Comment by hannestschofenig** [comment]
We now have to wait for Martin, Thomas, and Ekr to agree that the proposed changes are inline with their thinking and then I would submit a new version of the draft for the working group to look at.

File: draft-ietf-tls-dtls-connection-id.md [hunk 2]
```diff
@@ -134,34 +133,48 @@ indicates that the server will send with the client's CID but does not
...
 that allows new CIDs to be established mid-session, because
 DTLS 1.2 in general does not allow TLS 1.3-style post-handshake messages
-that do not themselves begin other handshakes. 
-
-DTLS peers switch to the new record layer format, i.e., the record layer format 
-containing the CID, when encryption is enabled.
-
-# Record Layer Extensions and Record Payload Protection
+that do not themselves begin other handshakes. When a DTLS session is 
+resumed, the "connection_id" extension is negotiated afresh.
```
Comment (martinthomson) [inline draft-ietf-tls-dtls-connection-id.md]: What about renegotiation?

File: draft-ietf-tls-dtls-connection-id.md [hunk 3]
```diff
@@ -134,34 +133,48 @@ indicates that the server will send with the client's CID but does not
...
+
+If DTLS peers have negotiated the use of a CIDs using the ClientHello and
+the ServerHello messages then the peers need to take two steps: 
+
+1. First, the peers need to determine whether incoming and ongoing messages need 
+to use the new record layer format, i.e., the record layer format 
+containing the CID. This new record layer format is only used once encryption 
+is enabled. Plaintext payloads never contain a CID in the record layer.
+Whenever a zero-length CID has been negotiated then the RFC 6347-defined 
+record format MUST be used (see Section 4.1 of {{RFC6347}}). For example,
```
Comment (martinthomson) [inline draft-ietf-tls-dtls-connection-id.md]: 6347 record format AND content type

File: draft-ietf-tls-dtls-connection-id.md [hunk 4]
```diff
@@ -134,34 +133,48 @@ indicates that the server will send with the client's CID but does not
...
+Whenever a zero-length CID has been negotiated then the RFC 6347-defined 
+record format MUST be used (see Section 4.1 of {{RFC6347}}). For example, 
+in {{dtls-example2}} the use of a CID has been successfully 
+negotiated between a client and a server whereby the client uses 
+the record format defined in this specification to transmit the CID value 
+(100) to the server. The server still uses the RFC 6347-defined 
+record format without a CID when transmitting records to the client. 
+
+2. Second, peers need to use the new MAC calculation defined in this document. 
+Any record layer payload that uses the new record layer format MUST use the
```
Comment (martinthomson) [inline draft-ietf-tls-dtls-connection-id.md]: And content type

File: draft-ietf-tls-dtls-connection-id.md [hunk 5]
```diff
@@ -171,7 +184,7 @@ connections which have negotiated this extension use a distinguished
 record type tls12_cid(25). Use of this content type has the following
 two implications:
 
-- The CID field is present
+- The CID field is present in the record
 - The true content type is inside the encryption envelope, as described
   below.
```
Comment (martinthomson) [inline draft-ietf-tls-dtls-connection-id.md:198]: + - The MAC calculation follows the process described in {{}}.

**Review by martinthomson — APPROVED** [review]
So I would slightly prefer to have zero- length CID mean use of the new content type, with the associated privacy enhancement, but this is fine.

**Comment by hannestschofenig** [comment]
Martin wrote: 

> I would slightly prefer to have zero- length CID mean use of the new content type, with the associated privacy enhancement

So proposal has not been made so far. We should discuss this idea. This is a new privacy-enhancing feature that does not necessarily have anything to do with the CID idea. In some sense this would allow the client and the server to negotiate the use of a traffic obfuscation technique.

Achim, Ekr, Martin, etc. -- what do you think?

File: draft-ietf-tls-dtls-connection-id.md (hunk 2 above)
Comment (hannestschofenig) [reply draft-ietf-tls-dtls-connection-id.md]: Good question. We don't use renegotiation in IoT deployments. It would be interesting to hear from someone who uses renegotation in the IoT context about what the expected behavior should be.

**Comment by boaks** [comment]
> So proposal has not been made so far.

That's the discussions in issues #22 and #43  
I linked to that issue #43 in the previous PR #35 as "Alternative III"
https://github.com/tlswg/dtls-conn-id/pull/35#issuecomment-460030833

To follow up the discussion is really difficult and time intensive. There is also one change left here
https://github.com/tlswg/dtls-conn-id/issues/26#issuecomment-459706334

That request to use the first CID not in FINISH and instead in the record after that.

**Comment by boaks** [comment]
> This is a new privacy-enhancing feature that does not necessarily have anything to do with the CID idea. 

I agree, this adds a new use-case to this draft and is not related to the CID idea.
It's related to the usage of the new introduced record type.

I created a issue about the scope of this draft #40 , which mentions that additional use-case. 
In the meantime I closed that scope issue. My current impression is, that the discussions are split and really hard to follow. Therefore I closed my additional issues in hope, that focus on the CID idea makes it possible to "release" this draft sooner.

**Comment by hannestschofenig** [comment]
You are correct. The proposal has been made before. I was confused....

**Comment by hannestschofenig** [comment]
Achim, do you prefer to use "alternative III" instead of what I wrote-up today?

**Comment by hannestschofenig** [comment]
Btw, I added notes to the draft about the additional privacy features. So, I believe I captured the scope issue you raised.

File: draft-ietf-tls-dtls-connection-id.md (hunk 3 above)
Comment (hannestschofenig) [reply draft-ietf-tls-dtls-connection-id.md]: Updated text

File: draft-ietf-tls-dtls-connection-id.md (hunk 4 above)
Comment (hannestschofenig) [reply draft-ietf-tls-dtls-connection-id.md]: updated text

File: draft-ietf-tls-dtls-connection-id.md (hunk 5 above)
Comment (hannestschofenig) [reply draft-ietf-tls-dtls-connection-id.md:198]: updated text

**Comment by boaks** [comment]
> Achim, do you prefer to use "alternative III" instead of what I wrote-up today?

I prefer any of the solutions! The difference between the flavours is for me much smaller, than the difference in shifting the timelines would be.

> Ekr, Martin, etc. -- 

For me, Hannes did a really great job. 
It would be also great, if you can give your "final" votes to give him a chance to finalize his work :-).

File: draft-ietf-tls-dtls-connection-id.md [hunk 6]
```diff
@@ -376,9 +392,9 @@ Importantly, the sequence number makes it possible for a passive attacker
 to correlate packets across CID changes. Thus, even if a client/server pair
 do a rehandshake to change CID, that does not provide much privacy benefit.
 
-This document does not change the security properties of DTLS {{RFC6347}}.
-It merely provides a more robust mechanism for associating an incoming packet
-with a stored security context.
+The CID-enhanced record layer introduces record padding; a privacy feature 
+not available with the original DTLS 1.2 RFC. Padding allows to inflate the 
+size of the ciphertext making traffic analysis more difficult.
```
Comment (boaks) [inline draft-ietf-tls-dtls-connection-id.md]: Just to mention:
For that feature it would make sense, to use the new record type even if the CID is empty.
That's what the issues #43 and the discussion in #22 is about.

As already commented, I have no preference.

**Comment by ekr** [comment]
I can live with that proposal as well, as long as we agree that it's a
violation to send an empty CID when one is expected.

**Comment by boaks** [comment]
> I can live with that proposal as well, as long as we agree that it's a violation to send an empty CID when one is expected.

The idea was to use the new record type even if an empty CID is negotiated. 
If a none empty CID is negotiated, then this none empty CID has to be used. 
The mentioned "violation" was, at least from my understand, not intended.

File: draft-ietf-tls-dtls-connection-id.md [hunk 7]
```diff
@@ -134,44 +133,59 @@ indicates that the server will send with the client's CID but does not
...
-containing the CID, when encryption is enabled.
-
-# Record Layer Extensions and Record Payload Protection
+that do not themselves begin other handshakes. When a DTLS session is 
+resumed, the "connection_id" extension is negotiated afresh. 
+
+If DTLS peers have negotiated the use of a CIDs using the ClientHello and
+the ServerHello messages then the peers need to take two steps: 
+
+1. First, the peers need to determine whether incoming and ongoing messages need
```
Comment (thomas-fossati) [inline draft-ietf-tls-dtls-connection-id.md]: s/ongoing/outgoing/

File: draft-ietf-tls-dtls-connection-id.md [hunk 8]
```diff
@@ -134,44 +133,59 @@ indicates that the server will send with the client's CID but does not
...
-# Record Layer Extensions and Record Payload Protection
+that do not themselves begin other handshakes. When a DTLS session is 
+resumed, the "connection_id" extension is negotiated afresh. 
+
+If DTLS peers have negotiated the use of a CIDs using the ClientHello and
+the ServerHello messages then the peers need to take two steps: 
+
+1. First, the peers need to determine whether incoming and ongoing messages need 
+to use the new record layer format, i.e., the record layer format 
+containing the CID. This new record layer format is only used once encryption
```
Comment (thomas-fossati) [inline draft-ietf-tls-dtls-connection-id.md]: To be sure I'd make this a MUST (and the following a MUST NOT) as this is necessary to interop.

I'd also add some more precision about the exact message where we expect the first CID to appear.

File: draft-ietf-tls-dtls-connection-id.md (hunk 2 above)
Comment (hannestschofenig) [reply draft-ietf-tls-dtls-connection-id.md]: What about saying for now "When a DTLS session is resumed or renegotiated, the "connection_id" extension is negotiated afresh."?

File: draft-ietf-tls-dtls-connection-id.md (hunk 7 above)
Comment (hannestschofenig) [reply draft-ietf-tls-dtls-connection-id.md]: OK. Fixed.

**Comment by boaks** [comment]
Fine! 
AFAIK all discussion about features/scope are concluded with that!
FMPOV #43, #23. #22 are fixed, when this PR is merged.

#18 will then be the only open issue, asking for implementations :-).

**Comment by boaks** [comment]
Uups, I think, I got a very late catch:

https://github.com/tlswg/dtls-conn-id/blob/master/draft-ietf-tls-dtls-connection-id.md#record-layer-extensions-and-record-payload-protection

> cid_length : 1 byte field indicating the length of the negotiated CID. If a zero-length CID has been negotiated, and therefore no CID appears on the wire, a cid_length of zero (0) MUST be added.

**If a zero-length CID has been negotiated, and therefore no CID appears on the wire, a cid_length of zero (0) MUST be added.** 

Is not longer valid!

For a empty CID, the RFC6347 record is used with the RFC6347 MAC and not the new MAC.

File: draft-ietf-tls-dtls-connection-id.md [hunk 9]
```diff
@@ -134,44 +133,67 @@ indicates that the server will send with the client's CID but does not
...
 In DTLS 1.2, CIDs are exchanged at the beginning of the DTLS
 session only. There is no dedicated "CID update" message
 that allows new CIDs to be established mid-session, because
 DTLS 1.2 in general does not allow TLS 1.3-style post-handshake messages
-that do not themselves begin other handshakes. 
+that do not themselves begin other handshakes. When a DTLS session is 
+resumed or renegotiated, the "connection_id" extension is negotiated afresh. 
+
+If DTLS peers have not negotiated the use of a CIDs then the RFC 6347-defined 
+record format and content type MUST be used.
```
Comment (martinthomson) [inline draft-ietf-tls-dtls-connection-id.md:155]: Add the zero-length CID thing here.

File: draft-ietf-tls-dtls-connection-id.md [hunk 10]
```diff
@@ -134,44 +133,67 @@ indicates that the server will send with the client's CID but does not
...
 that allows new CIDs to be established mid-session, because
 DTLS 1.2 in general does not allow TLS 1.3-style post-handshake messages
-that do not themselves begin other handshakes. 
+that do not themselves begin other handshakes. When a DTLS session is 
+resumed or renegotiated, the "connection_id" extension is negotiated afresh. 
+
+If DTLS peers have not negotiated the use of a CIDs then the RFC 6347-defined 
+record format and content type MUST be used. 
+
+If DTLS peers have negotiated the use of a CIDs using the ClientHello and
```
Comment (martinthomson) [inline draft-ietf-tls-dtls-connection-id.md:157]: Add non-zero-length here.

File: draft-ietf-tls-dtls-connection-id.md [hunk 11]
```diff
@@ -134,44 +133,67 @@ indicates that the server will send with the client's CID but does not
...
+the ServerHello messages then the peers need to take the following steps.
 
-DTLS peers switch to the new record layer format, i.e., the record layer format 
-containing the CID, when encryption is enabled.
+The DTLS peers determine whether incoming and outgoing messages need 
+to use the new record format, i.e., the record format containing the CID.
```
Comment (martinthomson) [inline draft-ietf-tls-dtls-connection-id.md:161]: With the above adjustments, this is simply: DTLS peers use the new content type, record format, and MAC calculation for all protected records.

File: draft-ietf-tls-dtls-connection-id.md [hunk 12]
```diff
@@ -134,44 +133,67 @@ indicates that the server will send with the client's CID but does not
...
-DTLS peers switch to the new record layer format, i.e., the record layer format 
-containing the CID, when encryption is enabled.
+The DTLS peers determine whether incoming and outgoing messages need 
+to use the new record format, i.e., the record format containing the CID. 
+The new record format and the CID content type is only used once encryption 
+is enabled. Plaintext payloads never use the new record type and the CID content 
+type. 
 
-# Record Layer Extensions and Record Payload Protection
+For sending, if a zero-length CID has been negotiated then the RFC 6347-defined
```
Comment (martinthomson) [inline draft-ietf-tls-dtls-connection-id.md:166]: Why limit this with "for sending"?

File: draft-ietf-tls-dtls-connection-id.md [hunk 13]
```diff
@@ -134,44 +133,67 @@ indicates that the server will send with the client's CID but does not
...
 {{I-D.ietf-tls-dtls13}} specifies how to carry the CID in DTLS 1.3.
 
 In order to allow a receiver to determine whether a record has CID or not,
 connections which have negotiated this extension use a distinguished
 record type tls12_cid(25). Use of this content type has the following
-two implications:
+three implications:
 
-- The CID field is present
+- The CID field is present.
```
Comment (martinthomson) [inline draft-ietf-tls-dtls-connection-id.md]:
```suggestion
- The CID field is present and contains one or more bytes.
```

File: draft-ietf-tls-dtls-connection-id.md (hunk 11 above)
Comment (boaks) [reply draft-ietf-tls-dtls-connection-id.md:161]: Please explain "all protected records". 
A RFC6347 record may be considered to be protected as well.

Comment (martinthomson) [reply draft-ietf-tls-dtls-connection-id.md:161]: Pay special attention to the context.  Which is when a non-zero-length CID is negotiated.

File: draft-ietf-tls-dtls-connection-id.md (hunk 12 above)
Comment (hannestschofenig) [reply draft-ietf-tls-dtls-connection-id.md:166]: I wrote the text in a way that it talks about sending and receiving. This is just a writing style.

File: draft-ietf-tls-dtls-connection-id.md (hunk 10 above)
Comment (hannestschofenig) [reply draft-ietf-tls-dtls-connection-id.md:157]: I don't think I need to add extra text at this point in the description since I talk about the non-zero length case a few sentences later.

File: draft-ietf-tls-dtls-connection-id.md (hunk 11 above)
Comment (hannestschofenig) [reply draft-ietf-tls-dtls-connection-id.md:161]: We could try to shorten the text but currently the style is similar to how Ekr described our conclusion in his email. I liked the way of separating the sending and the receiving processing description.

File: draft-ietf-tls-dtls-connection-id.md (hunk 6 above)
Comment (hannestschofenig) [reply draft-ietf-tls-dtls-connection-id.md]: But we are not going to add extra functionality in DTLS 1.2 when we already have the same features in 1.3

File: draft-ietf-tls-dtls-connection-id.md [hunk 14]
```diff
@@ -97,8 +97,7 @@ The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
 "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this
 document are to be interpreted as described in RFC 2119 {{RFC2119}}.
 
-The reader is assumed to be familiar with DTLS {{RFC6347}}.
-
+The reader is assumed to be familiar with DTLS 1.2 {{RFC6347}}.
```
Comment (chris-wood) [inline draft-ietf-tls-dtls-connection-id.md]: nit: perhaps rewrite to "This document assumes familiarity with DTLS 1.2 {{RFC6347}}."?

File: draft-ietf-tls-dtls-connection-id.md [hunk 15]
```diff
@@ -134,44 +133,67 @@ indicates that the server will send with the client's CID but does not
...
+When transmitting a datagram with the new record format and the new content type, 
+the new MAC computation defined in {{mac}} MUST be used.
+
+For receiving, if the tls12_cid content type is set, then the CID is used to look up 
+the connection and the security association. If the tls12_cid content type is not set, 
+then the connection and security association is looked up by the 5-tuple and a 
+check MUST be made to determine whether the expected CID value is indeed 
+zero length. If the check fails, then the datagram MUST be dropped. 
+
+When receiving a datagram with the new record format and the new content type,
```
Comment (chris-wood) [inline draft-ietf-tls-dtls-connection-id.md]: nit: s/and the new content type/and tls12_cid content type

File: draft-ietf-tls-dtls-connection-id.md [hunk 16]
```diff
@@ -134,44 +133,67 @@ indicates that the server will send with the client's CID but does not
...
+else the new record layer format and the new content type defined in {{dtls-record12}} MUST be used. 
+
+When transmitting a datagram with the new record format and the new content type, 
+the new MAC computation defined in {{mac}} MUST be used.
+
+For receiving, if the tls12_cid content type is set, then the CID is used to look up 
+the connection and the security association. If the tls12_cid content type is not set, 
+then the connection and security association is looked up by the 5-tuple and a 
+check MUST be made to determine whether the expected CID value is indeed 
+zero length. If the check fails, then the datagram MUST be dropped.
```
Comment (chris-wood) [inline draft-ietf-tls-dtls-connection-id.md:177]: Rather than check for a zero length CID, should we say that the recipient datagram MUST adhere to the format in RFC6347, else be dropped?

File: draft-ietf-tls-dtls-connection-id.md (hunk 15 above)
Comment (chris-wood) [inline draft-ietf-tls-dtls-connection-id.md]: When reading this condition one might assume that it's possible to receive a datagram with the new record format but without the new content type. Should we instead fold this together and say that (a) if the new content type is used then the new record format MUST also be in effect, and (b) if the new content type is not used then the old record format MUST be used?

File: draft-ietf-tls-dtls-connection-id.md [hunk 17]
```diff
@@ -376,9 +397,9 @@ Importantly, the sequence number makes it possible for a passive attacker
 to correlate packets across CID changes. Thus, even if a client/server pair
 do a rehandshake to change CID, that does not provide much privacy benefit.
 
-This document does not change the security properties of DTLS {{RFC6347}}.
-It merely provides a more robust mechanism for associating an incoming packet
-with a stored security context.
+The CID-enhanced record layer introduces record padding; a privacy feature 
+not available with the original DTLS 1.2 RFC. Padding allows to inflate the 
+size of the ciphertext making traffic analysis more difficult.
```
Comment (chris-wood) [inline draft-ietf-tls-dtls-connection-id.md]: Should we reference RFC8446 for more details about this padding and its traffic analysis implications?

**Review by chris-wood — CHANGES_REQUESTED** [review]

File: draft-ietf-tls-dtls-connection-id.md (hunk 16 above)
Comment (boaks) [reply draft-ietf-tls-dtls-connection-id.md:177]: The check is done against the associated "connection and security association" (the negotiated CID), not the received record.

Comment (chris-wood) [reply draft-ietf-tls-dtls-connection-id.md:177]: Ah! Thanks for the clarification.

File: draft-ietf-tls-dtls-connection-id.md (hunk 17 above)
Comment (hannestschofenig) [reply draft-ietf-tls-dtls-connection-id.md]: Yes. A reference to RFC8446 makes sense here.

File: draft-ietf-tls-dtls-connection-id.md (hunk 14 above)
Comment (hannestschofenig) [reply draft-ietf-tls-dtls-connection-id.md]: Sounds good.

File: draft-ietf-tls-dtls-connection-id.md (hunk 15 above)
Comment (hannestschofenig) [reply draft-ietf-tls-dtls-connection-id.md]: Fixed.

Comment (hannestschofenig) [reply draft-ietf-tls-dtls-connection-id.md]: I tried to change the wording to make it more clear that the new record format and the new content type is actually the same. There is no new record format that does not contain the new content type. 

Have a look at the new text

**Review by chris-wood — APPROVED** [review]
